### PR TITLE
Refine torch audio pipeline tests: adding frame splicing, fix sequence length calculation, reflect pad start/end of the signal

### PR DESCRIPTION
**Pull request by jantonguirao**
Signed-off-by: Joaquin Anton <janton@nvidia.com>

#### Why we need this PR?
*Pick one, remove the rest*
- It adds tests for spectrogram frame splicing

#### What happened in this PR?
*Fill relevant points, put NA otherwise. Replace anything inside []*
 - What solution was applied:
     *Extended torch audio pipeline tests to cover the frame splicing operation*
 - Affected modules and functionalities:
     *torch audio pipeline tests*
 - Key points relevant for the review:
     *NA*
 - Validation and testing:
     *New tests*
 - Documentation (including examples):
     *NA*

**JIRA TASK**: *[DALI-2112]*

### CODE REVIEW

File: dali/test/python/test_torch_pipeline_rnnt.py [hunk 1]
```diff
@@ -355,77 +354,161 @@ def test_torch_vs_dali_normalize():
         for normalize_type in ['per_feature', 'all_features']:
             yield _testimpl_torch_vs_dali_normalize, normalize_type, device
 
+def torch_frame_splicing(mel_spec, stacking=1, subsampling=1, device='cpu'):
+    mel_spec = torch.tensor(mel_spec, dtype=torch.float32).unsqueeze(0)
+    if device == 'gpu':
+        mel_spec = mel_spec.cuda()
+    out = stack_subsample_frames(mel_spec, stacking=stacking, subsampling=subsampling)
+    out = out.cpu().detach().numpy().squeeze(0)
```
Comment (JanuszL) [inline dali/test/python/test_torch_pipeline_rnnt.py:362]: Why you need to transfer it to the CPU and use numpy squeeze?
Can't you leave it at the GPU and use torch squeeze?

File: dali/test/python/test_torch_pipeline_rnnt.py [hunk 2]
```diff
@@ -355,77 +354,161 @@ def test_torch_vs_dali_normalize():
...
-        log_features = fn.transpose(log_features, perm=[1, 0])
-        log_features = fn.reshape(log_features, rel_shape=[-1, frame_splicing])
-        log_features = fn.pad(log_features, axes=[0], fill_value=0, align=frame_splicing, shape=[1])
-        log_features = fn.transpose(log_features, perm=[1, 0])
+        seq = [log_features]
+        for n in range(1, frame_splicing):
+            f = fn.slice(log_features, start=n, axes=(1,))
+            f = fn.pad(f, shape=fn.cat(nfeatures, spec_len))
+            seq.append(f)
+        log_features_spliced = fn.cat(*seq, axis=0)
```
Comment (JanuszL) [inline dali/test/python/test_torch_pipeline_rnnt.py:436]: I wonder how much of the GPU memory it would consume (as we add multiple slice and pad operators)?
When we have better memory management the buffer probably could be reused between instances, but not now.

File: dali/test/python/test_torch_pipeline_rnnt.py [hunk 3]
```diff
@@ -355,77 +354,161 @@ def test_torch_vs_dali_normalize():
...
-        preemph_audio = np.array(dali_out[4][0])
-        audio = np.array(dali_out[5][0])
-        ref = reference_data[i].squeeze(0)
-        assert ref.shape == norm_log_features.shape
+        log_features_spliced = np.array(dali_out[1][0])
+        log_features = np.array(dali_out[2][0])
+        mel_spec = np.array(dali_out[3][0])
+        spec = np.array(dali_out[4][0])
+        preemph_audio = np.array(dali_out[5][0])
+        audio = np.array(dali_out[6][0])
```
Comment (JanuszL) [inline dali/test/python/test_torch_pipeline_rnnt.py:496]: How about using `to_array` from test_utils.py.

File: dali/test/python/test_torch_pipeline_rnnt.py (hunk 1 above)
Comment (jantonguirao) [reply dali/test/python/test_torch_pipeline_rnnt.py:362]: Only for convenience. This functions process a single sample, the input is a numpy array and the output as well. This way I can compare torch and DALI implementations side by side. This function is meant as a utility to test every step of the pipeline separately.

File: dali/test/python/test_torch_pipeline_rnnt.py (hunk 2 above)
Comment (jantonguirao) [reply dali/test/python/test_torch_pipeline_rnnt.py:436]: Now that I think of it, I don't need the pad. I can simply use slice with out-of-bounds policy set to pad.

File: dali/test/python/test_torch_pipeline_rnnt.py (hunk 3 above)
Comment (jantonguirao) [reply dali/test/python/test_torch_pipeline_rnnt.py:496]: I will.

File: dali/test/python/test_torch_pipeline_rnnt.py (hunk 1 above)
Comment (jantonguirao) [reply dali/test/python/test_torch_pipeline_rnnt.py:362]: I see that detach is not here. I'll remove.
squeeze(0) is there because the torch function work with a batch and I want to process a single sample, so I unsqueeze(0) before feeding the function and then squeeze(0) the result to create the batch dimension.

File: dali/test/python/test_torch_pipeline_rnnt.py [hunk 4]
```diff
@@ -112,8 +112,7 @@ def normalize_batch(x, seq_len, normalize_type):
             return x
 
     def get_seq_len(self, seq_len):
-        return torch.ceil(seq_len.to(dtype=torch.float) / self.hop_length).to(
-            dtype=torch.int)
+        return seq_len.to(dtype=torch.int) // self.hop_length + 1
```
Comment (jantonguirao) [inline dali/test/python/test_torch_pipeline_rnnt.py:115]: This is a bug in the reference.

Comment (mzient) [reply dali/test/python/test_torch_pipeline_rnnt.py:115]: I don't understand. The new code is a bug or the old one? This one looks broken.

**Review by JanuszL — APPROVED** [review]

File: dali/test/python/test_torch_pipeline_rnnt.py [hunk 5]
```diff
@@ -182,7 +181,7 @@ def torch_spectrogram(audio, sample_rate, device='cpu',
                           center=center, window=window_tensor.to(dtype=torch.float))
     # get power spectrum
     spectrogram = stft_out.pow(2).sum(-1)
-    spectrogram = spectrogram.cpu().detach().numpy()
+    spectrogram = spectrogram.cpu().numpy()
```
Comment (mzient) [inline dali/test/python/test_torch_pipeline_rnnt.py:184]: Why did you remove `detach` everywhere?

File: dali/test/python/test_torch_pipeline_rnnt.py [hunk 6]
```diff
@@ -355,77 +354,149 @@ def test_torch_vs_dali_normalize():
...
 
     if frame_splicing > 1:
-        log_features = fn.transpose(log_features, perm=[1, 0])
-        log_features = fn.reshape(log_features, rel_shape=[-1, frame_splicing])
-        log_features = fn.pad(log_features, axes=[0], fill_value=0, align=frame_splicing, shape=[1])
-        log_features = fn.transpose(log_features, perm=[1, 0])
+        seq = [log_features]
+        for n in range(1, frame_splicing):
+            f = fn.slice(log_features, start=n, shape=spec_len, axes=(1,), out_of_bounds_policy='pad', fill_values=0)
+            seq.append(f)
```
Comment (mzient) [inline dali/test/python/test_torch_pipeline_rnnt.py:433]: No subsampling?

File: dali/test/python/test_torch_pipeline_rnnt.py [hunk 7]
```diff
@@ -355,77 +354,149 @@ def test_torch_vs_dali_normalize():
...
+                f = fn.slice(x, start=n, shape=in_len, axes=(1,), out_of_bounds_policy='pad', fill_values=0)
+                seq.append(f)
+            x = fn.cat(*seq, axis=0)
+            nfeatures = nfeatures * stacking
+        if subsampling > 1:
+            out_len = (in_len + subsampling - 1) // subsampling
+            m = fn.transforms.scale(scale=[subsampling, 1], center=[0.5, 0])
+            x = fn.reshape(x, rel_shape=[1, 1, -1], layout="HWC")  # Layout required by WarpAffine
+            x = fn.warp_affine(x, matrix=m, size=fn.cat(nfeatures, out_len), interp_type=types.INTERP_NN)
+            x = fn.reshape(x, rel_shape=[1, 1], layout="ft")
```
Comment (mzient) [inline dali/test/python/test_torch_pipeline_rnnt.py:385]: I think this logic should be extracted to a function and used in the full pipeline.

Comment (jantonguirao) [reply dali/test/python/test_torch_pipeline_rnnt.py:385]: Done

File: dali/test/python/test_torch_pipeline_rnnt.py (hunk 6 above)
Comment (jantonguirao) [reply dali/test/python/test_torch_pipeline_rnnt.py]: Added it.

File: dali/test/python/test_torch_pipeline_rnnt.py (hunk 5 above)
Comment (jantonguirao) [reply dali/test/python/test_torch_pipeline_rnnt.py:193]: not necessary, I think.

File: dali/test/python/test_torch_pipeline_rnnt.py (hunk 4 above)
Comment (jantonguirao) [reply dali/test/python/test_torch_pipeline_rnnt.py:118]: The old one is the bug. The new one is the actual length of the spectrogram.

**Review by mzient — APPROVED** [review]

**Comment by jantonguirao** [comment]
!build
